File: src/utils/curseforge-api.zig
const std = @import("std");
const httpRequest = @import("http-request.zig");
const decompress = @import("decompress.zig");

/// CurseForge API response structure based on the Get Files Response schema
/// Reference: https://docs.curseforge.com/rest-api/#tocS_Get%20Files%20Response
pub const GetFilesResponse = struct {
    data: []FileData,
    pagination: ?Pagination = null,

    pub const FileData = struct {
        id: i32,
        gameId: i32,
        modId: i32,
        isAvailable: bool,
        displayName: []const u8,
        fileName: []const u8,
        releaseType: i32,
        fileStatus: i32,
        hashes: []FileHash,
        fileDate: []const u8,
        fileLength: i64,
        downloadCount: i64,
        fileSizeOnDisk: ?i64 = null,
        downloadUrl: ?[]const u8 = null,
        gameVersions: [][]const u8,
        sortableGameVersions: []SortableGameVersion,
        dependencies: []FileDependency,
        exposeAsAlternative: ?bool = null,
        parentProjectFileId: ?i32 = null,
        alternateFileId: ?i32 = null,
        isServerPack: ?bool = null,
        serverPackFileId: ?i32 = null,
        isEarlyAccessContent: ?bool = null,
        earlyAccessEndDate: ?[]const u8 = null,
        fileFingerprint: i64,
        modules: []FileModule,

        pub const FileHash = struct {
            value: []const u8,
            algo: i32,
        };

        pub const SortableGameVersion = struct {
            gameVersionName: []const u8,
            gameVersionPadded: []const u8,
            gameVersion: []const u8,
            gameVersionReleaseDate: []const u8,
            gameVersionTypeId: ?i32 = null,
        };

        pub const FileDependency = struct {
            modId: i32,
            relationType: i32,
        };

        pub const FileModule = struct {
            name: []const u8,
            fingerprint: i64,
        };
    };

    pub const Pagination = struct {
        index: i32,
        pageSize: i32,
        resultCount: i32,
        totalCount: i64,
    };
};

/// Parses CurseForge API response JSON into GetFilesResponse struct
///
/// Args:
///   allocator: Memory allocator for parsing
///   jsonData: Raw JSON response from CurseForge API
///
/// Returns:
///   Parsed GetFilesResponse structure
///
/// Errors:
///   - error.OutOfMemory: If allocation fails
///   - error.SyntaxError: If JSON is malformed
///   - error.UnexpectedToken: If JSON structure doesn't match expected schema
pub fn parseGetFilesResponse(allocator: std.mem.Allocator, jsonData: []const u8) !GetFilesResponse {
    const parsed = try std.json.parseFromSlice(GetFilesResponse, allocator, jsonData, .{
        .ignore_unknown_fields = true,
        .allocate = .alloc_always,
    });
    return parsed.value;
}

/// Extracts the download URL from the first file in the response
///
/// Args:
///   response: Parsed GetFilesResponse structure
///
/// Returns:
///   Download URL string or null if not found
pub fn extractDownloadUrl(response: *const GetFilesResponse) ?[]const u8 {
    if (response.data.len == 0) return null;
    return response.data[0].downloadUrl;
}

test "parseGetFilesResponse with minimal data" {
    const allocator = std.testing.allocator;

    const testJson =
        \\{
        \\  "data": [
        \\    {
        \\      "id": 12345,
        \\      "gameId": 432,
        \\      "modId": 67890,
        \\      "isAvailable": true,
        \\      "displayName": "Test Modpack",
        \\      "fileName": "test-modpack-1.0.0.zip",
        \\      "releaseType": 1,
        \\      "fileStatus": 4,
        \\      "hashes": [],
        \\      "fileDate": "2023-12-01T10:00:00.000Z",
        \\      "fileLength": 1048576,
        \\      "downloadCount": 100,
        \\      "downloadUrl": "https://example.com/test-modpack.zip",
        \\      "gameVersions": ["1.21.1"],
        \\      "sortableGameVersions": [],
        \\      "dependencies": [],
        \\      "fileFingerprint": 123456789,
        \\      "modules": []
        \\    }
        \\  ]
        \\}
    ;

    const response = try parseGetFilesResponse(allocator, testJson);
    defer std.json.parseFree(GetFilesResponse, response, .{ .allocator = allocator });

    try std.testing.expect(response.data.len == 1);
    try std.testing.expect(response.data[0].id == 12345);
    try std.testing.expectEqualStrings("Test Modpack", response.data[0].displayName);
    try std.testing.expectEqualStrings("test-modpack-1.0.0.zip", response.data[0].fileName);

    const downloadUrl = extractDownloadUrl(&response);
    try std.testing.expect(downloadUrl != null);
    try std.testing.expectEqualStrings("https://example.com/test-modpack.zip", downloadUrl.?);
}

test "extractDownloadUrl with empty data" {
    const response = GetFilesResponse{
        .data = &[_]GetFilesResponse.FileData{},
    };

    const downloadUrl = extractDownloadUrl(&response);
    try std.testing.expect(downloadUrl == null);
}

/// CurseForge API client for fetching modpack files
pub const CurseForgeClient = struct {
    allocator: std.mem.Allocator,
    httpClient: httpRequest,
    apiToken: []const u8,

    const CURSEFORGE_API_BASE_URL = "https://api.curseforge.com";

    pub fn init(allocator: std.mem.Allocator, apiToken: []const u8) !CurseForgeClient {
        const httpClient = try httpRequest.init(allocator);
        return CurseForgeClient{
            .allocator = allocator,
            .httpClient = httpClient,
            .apiToken = apiToken,
        };
    }

    pub fn deinit(self: *CurseForgeClient) void {
        self.httpClient.deinit();
    }

    /// Fetches the latest modpack file from CurseForge API
    ///
    /// Args:
    ///   modpackId: CurseForge modpack ID
    ///
    /// Returns:
    ///   GetFilesResponse containing modpack file information
    ///
    /// Errors:
    ///   - error.OutOfMemory: If allocation fails
    ///   - error.HttpRequestFailed: If HTTP request fails
    ///   - error.InvalidApiResponse: If API response is invalid
    pub fn getLatestModpackFile(self: *CurseForgeClient, modpackId: []const u8) !GetFilesResponse {
        // Build API URL
        const url = try std.fmt.allocPrint(self.allocator, "{s}/v1/mods/{s}/files?pageIndex=0&pageSize=1&sort=dateCreated&sortDescending=true&removeAlphas=true", .{ CURSEFORGE_API_BASE_URL, modpackId });
        defer self.allocator.free(url);

        // Prepare headers with API key
        const apiKeyHeader = try std.fmt.allocPrint(self.allocator, "{s}", .{self.apiToken});
        defer self.allocator.free(apiKeyHeader);

        var headers = [_]std.http.Header{
            .{ .name = "x-api-key", .value = apiKeyHeader },
            .{ .name = "Accept", .value = "application/json" },
        };

        // Make HTTP request
        const response = self.httpClient.get(url, headers[0..]) catch |err| {
            std.log.err("Failed to fetch modpack files from CurseForge API: {any}", .{err});
            return error.HttpRequestFailed;
        };
        defer self.allocator.free(response.body);

        if (response.status != .ok) {
            std.log.err("CurseForge API returned HTTP status: {any}", .{response.status});
            return error.HttpRequestFailed;
        }

        const decompressedResponse = if (decompress.isGzipCompressed(response.body))
            decompress.decompressGzip(std.heap.page_allocator, response.body) catch |err| {
                std.debug.print("Decompression failed: ✗ ({any})\n", .{err});
                return error.DecompressionFailed;
            }
        else
            try std.heap.page_allocator.dupe(u8, response.body);
        defer std.heap.page_allocator.free(decompressedResponse);

        // Parse JSON response
        const parsedResponse = parseGetFilesResponse(self.allocator, decompressedResponse) catch |err| {
            std.log.err("Failed to parse CurseForge API response: {any}", .{err});
            return error.InvalidApiResponse;
        };

        return parsedResponse;
    }

    /// Downloads modpack file from the given URL
    ///
    /// Args:
    ///   downloadUrl: URL to download the modpack file from
    ///
    /// Returns:
    ///   Raw bytes of the downloaded modpack file
    ///
    /// Errors:
    ///   - error.OutOfMemory: If allocation fails
    ///   - error.HttpRequestFailed: If HTTP request fails
    pub fn downloadModpackFile(self: *CurseForgeClient, downloadUrl: []const u8) ![]u8 {
        const response = self.httpClient.get(downloadUrl, &.{}) catch |err| {
            std.log.err("Failed to download modpack file from URL: {s}, error: {any}", .{ downloadUrl, err });
            return error.HttpRequestFailed;
        };

        if (response.status != .ok) {
            std.log.err("Download request returned HTTP status: {any}", .{response.status});
            self.allocator.free(response.body);
            return error.HttpRequestFailed;
        }

        // Create a mutable copy of the response body for the caller
        const mutableCopy = try self.allocator.dupe(u8, response.body);
        self.allocator.free(response.body);
        return mutableCopy;
    }
};

test "CurseForgeClient initialization" {
    const allocator = std.testing.allocator;
    const testApiToken = "test-api-token";

    var client = try CurseForgeClient.init(allocator, testApiToken);
    defer client.deinit();

    try std.testing.expectEqualStrings(testApiToken, client.apiToken);
}

File: src/utils/curseforge-api.py
from __future__ import annotations

import io
from dataclasses import dataclass, field
from typing import Any, Dict, Iterable, List, Optional

import requests

from constants import DEFAULT_HEADERS, DOWNLOAD_BUFFER_SIZE, HTTP_TIMEOUT
from errors import CurseForgeApiError, ModpackDownloadError


CURSEFORGE_API_BASE_URL = "https://api.curseforge.com"


def _parseInt(value: Any) -> Optional[int]:
    if isinstance(value, bool):
        return int(value)
    if isinstance(value, (int, float)):
        return int(value)
    if isinstance(value, str):
        stripped = value.strip()
        if not stripped:
            return None
        try:
            return int(stripped)
        except ValueError:
            return None
    return None


def _toInt(value: Any, default: int = 0) -> int:
    parsed = _parseInt(value)
    return parsed if parsed is not None else default


def _toOptionalInt(value: Any) -> Optional[int]:
    return _parseInt(value)


def _toOptionalBool(value: Any) -> Optional[bool]:
    if isinstance(value, bool):
        return value
    if isinstance(value, (int, float)):
        return bool(value)
    if isinstance(value, str):
        lowered = value.strip().lower()
        if lowered in {"true", "1", "yes"}:
            return True
        if lowered in {"false", "0", "no"}:
            return False
    return None


def _toStr(value: Any, default: str = "") -> str:
    if value is None:
        return default
    return str(value)


def _mapList(items: Iterable[Any], mapper):
    return [mapper(item) for item in items if isinstance(item, dict)]


@dataclass
class FileHash:
    value: str = ""
    algo: int = 0

    @classmethod
    def fromDict(cls, data: Dict[str, Any]) -> "FileHash":
        return cls(
            value=_toStr(data.get("value")),
            algo=_toInt(data.get("algo")),
        )


@dataclass
class SortableGameVersion:
    gameVersionName: str = ""
    gameVersionPadded: str = ""
    gameVersion: str = ""
    gameVersionReleaseDate: str = ""
    gameVersionTypeId: Optional[int] = None

    @classmethod
    def fromDict(cls, data: Dict[str, Any]) -> "SortableGameVersion":
        return cls(
            gameVersionName=_toStr(data.get("gameVersionName")),
            gameVersionPadded=_toStr(data.get("gameVersionPadded")),
            gameVersion=_toStr(data.get("gameVersion")),
            gameVersionReleaseDate=_toStr(data.get("gameVersionReleaseDate")),
            gameVersionTypeId=_toOptionalInt(data.get("gameVersionTypeId")),
        )


@dataclass
class FileDependency:
    modId: int = 0
    relationType: int = 0

    @classmethod
    def fromDict(cls, data: Dict[str, Any]) -> "FileDependency":
        return cls(
            modId=_toInt(data.get("modId")),
            relationType=_toInt(data.get("relationType")),
        )


@dataclass
class FileModule:
    name: str = ""
    fingerprint: int = 0

    @classmethod
    def fromDict(cls, data: Dict[str, Any]) -> "FileModule":
        return cls(
            name=_toStr(data.get("name")),
            fingerprint=_toInt(data.get("fingerprint")),
        )


@dataclass
class FileData:
    id: int = 0
    gameId: int = 0
    modId: int = 0
    isAvailable: bool = False
    displayName: str = ""
    fileName: str = ""
    releaseType: int = 0
    fileStatus: int = 0
    hashes: List[FileHash] = field(default_factory=list)
    fileDate: str = ""
    fileLength: int = 0
    downloadCount: int = 0
    fileSizeOnDisk: Optional[int] = None
    downloadUrl: Optional[str] = None
    gameVersions: List[str] = field(default_factory=list)
    sortableGameVersions: List[SortableGameVersion] = field(default_factory=list)
    dependencies: List[FileDependency] = field(default_factory=list)
    exposeAsAlternative: Optional[bool] = None
    parentProjectFileId: Optional[int] = None
    alternateFileId: Optional[int] = None
    isServerPack: Optional[bool] = None
    serverPackFileId: Optional[int] = None
    isEarlyAccessContent: Optional[bool] = None
    earlyAccessEndDate: Optional[str] = None
    fileFingerprint: int = 0
    modules: List[FileModule] = field(default_factory=list)

    @classmethod
    def fromDict(cls, data: Dict[str, Any]) -> "FileData":
        return cls(
            id=_toInt(data.get("id")),
            gameId=_toInt(data.get("gameId")),
            modId=_toInt(data.get("modId")),
            isAvailable=bool(data.get("isAvailable", False)),
            displayName=_toStr(data.get("displayName")),
            fileName=_toStr(data.get("fileName")),
            releaseType=_toInt(data.get("releaseType")),
            fileStatus=_toInt(data.get("fileStatus")),
            hashes=_mapList(data.get("hashes", []), FileHash.fromDict),
            fileDate=_toStr(data.get("fileDate")),
            fileLength=_toInt(data.get("fileLength")),
            downloadCount=_toInt(data.get("downloadCount")),
            fileSizeOnDisk=_toOptionalInt(data.get("fileSizeOnDisk")),
            downloadUrl=_toStr(data.get("downloadUrl")) if data.get("downloadUrl") is not None else None,
            gameVersions=[_toStr(item) for item in data.get("gameVersions", [])],
            sortableGameVersions=_mapList(data.get("sortableGameVersions", []), SortableGameVersion.fromDict),
            dependencies=_mapList(data.get("dependencies", []), FileDependency.fromDict),
            exposeAsAlternative=_toOptionalBool(data.get("exposeAsAlternative")),
            parentProjectFileId=_toOptionalInt(data.get("parentProjectFileId")),
            alternateFileId=_toOptionalInt(data.get("alternateFileId")),
            isServerPack=_toOptionalBool(data.get("isServerPack")),
            serverPackFileId=_toOptionalInt(data.get("serverPackFileId")),
            isEarlyAccessContent=_toOptionalBool(data.get("isEarlyAccessContent")),
            earlyAccessEndDate=_toStr(data.get("earlyAccessEndDate")) if data.get("earlyAccessEndDate") is not None else None,
            fileFingerprint=_toInt(data.get("fileFingerprint")),
            modules=_mapList(data.get("modules", []), FileModule.fromDict),
        )


@dataclass
class Pagination:
    index: int = 0
    pageSize: int = 0
    resultCount: int = 0
    totalCount: int = 0

    @classmethod
    def fromDict(cls, data: Dict[str, Any]) -> "Pagination":
        return cls(
            index=_toInt(data.get("index")),
            pageSize=_toInt(data.get("pageSize")),
            resultCount=_toInt(data.get("resultCount")),
            totalCount=_toInt(data.get("totalCount")),
        )


@dataclass
class GetFilesResponse:
    data: List[FileData] = field(default_factory=list)
    pagination: Optional[Pagination] = None


def parseGetFilesResponse(payload: Dict[str, Any]) -> GetFilesResponse:
    if not isinstance(payload, dict):
        raise CurseForgeApiError("CurseForge API response payload is not a JSON object")

    rawData = payload.get("data", [])
    if not isinstance(rawData, list):
        raise CurseForgeApiError("CurseForge API response payload is missing a 'data' array")

    files = [FileData.fromDict(item) for item in rawData if isinstance(item, dict)]

    paginationData = payload.get("pagination")
    pagination = Pagination.fromDict(paginationData) if isinstance(paginationData, dict) else None

    return GetFilesResponse(data=files, pagination=pagination)


def extractDownloadUrl(response: GetFilesResponse) -> Optional[str]:
    if not response.data:
        return None
    return response.data[0].downloadUrl


@dataclass
class CurseForgeClient:
    apiToken: str
    session: requests.Session = field(default_factory=requests.Session)

    def __post_init__(self) -> None:
        self.session.headers.update({
            "User-Agent": DEFAULT_HEADERS.get("User-Agent", "playtime-minecraft-bootstrap/py"),
        })

    def __enter__(self) -> "CurseForgeClient":
        return self

    def __exit__(self, exc_type, exc_value, traceback) -> None:
        self.close()

    def close(self) -> None:
        self.session.close()

    def getLatestModpackFile(self, modpackId: str) -> GetFilesResponse:
        url = (
            f"{CURSEFORGE_API_BASE_URL}/v1/mods/{modpackId}/files"
            "?pageIndex=0&pageSize=1&sort=dateCreated&sortDescending=true&removeAlphas=true"
        )

        headers = {
            "x-api-key": self.apiToken,
            "Accept": "application/json",
        }

        try:
            response = self.session.get(
                url,
                headers=headers,
                timeout=HTTP_TIMEOUT,
            )
        except requests.RequestException as exc:
            raise CurseForgeApiError(f"Failed to fetch modpack files from CurseForge API: {exc}") from exc

        if response.status_code >= 400:
            raise CurseForgeApiError(
                f"CurseForge API returned HTTP {response.status_code} for modpack '{modpackId}'"
            )

        try:
            payload = response.json()
        except ValueError as exc:
            raise CurseForgeApiError("CurseForge API response is not valid JSON") from exc

        return parseGetFilesResponse(payload)

    def downloadModpackFile(self, downloadUrl: str) -> bytes:
        headers = {
            "x-api-key": self.apiToken,
            "Accept": DEFAULT_HEADERS.get("Accept", "*/*"),
        }

        try:
            response = self.session.get(
                downloadUrl,
                headers=headers,
                timeout=HTTP_TIMEOUT,
                stream=True,
            )
        except requests.RequestException as exc:
            raise ModpackDownloadError(
                f"Failed to download modpack archive from '{downloadUrl}': {exc}"
            ) from exc

        if response.status_code >= 400:
            raise ModpackDownloadError(
                f"CurseForge download returned HTTP {response.status_code} for '{downloadUrl}'"
            )

        buffer = io.BytesIO()

        try:
            for chunk in response.iter_content(chunk_size=DOWNLOAD_BUFFER_SIZE):
                if chunk:
                    buffer.write(chunk)
        except requests.RequestException as exc:
            raise ModpackDownloadError(
                f"Error while streaming modpack archive from '{downloadUrl}': {exc}"
            ) from exc

        return buffer.getvalue()


